include/visitors/code_gen.h

```diff
@@ -2,30 +2,545 @@
+            return;
+        }
+
+        std::string filename = "output.wasm";
+        std::ofstream file(filename, std::ios::binary);
+        if (file.is_open())
+        {
+            file.write(static_cast<char *>(result.binary), result.binaryBytes);
+            file.close();
+            std::cout << "wasm module written to " << filename << std::endl;
+        }
+        else
+        {
+            std::cerr << "failed to open file" << filename << std::endl;
+        }
+
+        free(result.binary);
+
+        this->environment->pop_env();
+    }
+
+    int allocate_local()
+    {
+        int index = this->local_index++;
+        return index;
+    }
+
+    BinaryenType from_bird_type(Token token)
+    {
+        if (token.lexeme == "bool")
+            return BinaryenTypeInt32(); // no i1?
+        else if (token.lexeme == "int")
+            return BinaryenTypeInt32();
+        else if (token.lexeme == "float")
+            return BinaryenTypeFloat64();
+        else if (token.lexeme == "void")
+            return BinaryenTypeNone();
+        else if (token.lexeme == "str")
+            return BinaryenTypeInt32(); // ptr
+        else
+            throw BirdException("invalid type");
+    }
+
+    void visit_block(Block *block)
+    {
+        // throw BirdException("Implement visit_block");
+    }
+
+    void visit_decl_stmt(DeclStmt *decl_stmt)
+    {
+        decl_stmt->value->accept(this);
+        BinaryenExpressionRef initializer_value = this->stack.pop();
+
+        BinaryenType type = decl_stmt->type_identifier.has_value()
+                                ? from_bird_type(decl_stmt->type_identifier.value())
+                                : BinaryenExpressionGetType(initializer_value);
+
+        int index = allocate_local(); // locals defined by index instead of name
+        locals.push_back(type);
+
+        BinaryenExpressionRef set_local = BinaryenLocalSet(this->mod, index, initializer_value);
+
+        environment->declare(decl_stmt->identifier.lexeme, BinaryenLocalGet(this->mod, index, type));
+
+        fn_body.push_back(set_local);
+    }
+
+    void visit_assign_expr(AssignExpr *assign_expr)
+    {
+        // throw BirdException("Implement Assign Expression");
+    }
+
+    /*
+     * this doesnt work, not sure if i fully understand
+     * how to fill in the format strings. apparently there is no
+     * CreateGlobalString equivalent.
+     *
+     * INPUT:
+     * var x = 1;
+     * print x;
+     *
+     * OUTPUT:
+     * (module
+     *  (type $none_=>_none (func))
+     *  (type $i32_=>_none (func (param i32)))
+     *  (import "env" "printf" (func $printf (param i32)))
+     *  (export "main" (func $main))
+     *  (func $main
+     *   (local $0 i32)
+     *   (local.set $0
+     *    (i32.const 1)
+     *   )
+     *   (call $printf
+     *    (local.get $0)
+     *   )
+     *  )
+     * )
+     */
+    void visit_print_stmt(PrintStmt *print_stmt)
+    {
+        for (auto &arg : print_stmt->args)
```

Comment:
Let's wait on the print statement for now cause there are still a lot of unknowns here.